library/std/src/sys/unix/mutex.rs

```diff
@@ -54,14 +54,17 @@ impl Mutex {
         let mut attr = MaybeUninit::<libc::pthread_mutexattr_t>::uninit();
         cvt_nz(libc::pthread_mutexattr_init(attr.as_mut_ptr())).unwrap();
         let attr = PthreadMutexAttr(&mut attr);
-        cvt_nz(libc::pthread_mutexattr_settype(attr.0.as_mut_ptr(), libc::PTHREAD_MUTEX_NORMAL))
-            .unwrap();
+        cvt_nz(libc::pthread_mutexattr_settype(
+            attr.0.as_mut_ptr(),
+            libc::PTHREAD_MUTEX_ERRORCHECK,
+        ))
+        .unwrap();
         cvt_nz(libc::pthread_mutex_init(self.inner.get(), attr.0.as_ptr())).unwrap();
     }
     #[inline]
     pub unsafe fn lock(&self) {
         let r = libc::pthread_mutex_lock(self.inner.get());
-        debug_assert_eq!(r, 0);
+        assert_eq!(r, 0);
```

Comment:
This case is actually fairly well covered by `./src/test/ui/issues/issue-33770.rs` - but as a result of the new mutex type, an error check would be explicitly responsible for catching this, instead of an implied deadlock.

(I'd be happy to add an error message too, to indicate explicitly that a double-lock just occurred)